cmd/ingester/main.go

```diff
@@ -0,0 +1,176 @@
+	"github.com/spf13/viper"
+	"github.com/uber/jaeger-lib/metrics"
+	"go.uber.org/zap"
+
+	"github.com/jaegertracing/jaeger/cmd/env"
+	"github.com/jaegertracing/jaeger/cmd/flags"
+	"github.com/jaegertracing/jaeger/cmd/ingester/app"
+	spanconsumer "github.com/jaegertracing/jaeger/cmd/ingester/app/consumer"
+	"github.com/jaegertracing/jaeger/cmd/ingester/app/processor"
+	"github.com/jaegertracing/jaeger/pkg/config"
+	"github.com/jaegertracing/jaeger/pkg/kafka/consumer"
+	pMetrics "github.com/jaegertracing/jaeger/pkg/metrics"
+	"github.com/jaegertracing/jaeger/pkg/version"
+	"github.com/jaegertracing/jaeger/plugin/storage"
+	"github.com/jaegertracing/jaeger/storage/spanstore"
+)
+
+func main() {
+	var signalsChannel = make(chan os.Signal, 0)
+	signal.Notify(signalsChannel, os.Interrupt, syscall.SIGTERM)
+
+	storageFactory, err := storage.NewFactory(storage.FactoryConfigFromEnvAndCLI(os.Args, os.Stderr))
+	if err != nil {
+		log.Fatalf("Cannot initialize storage factory: %v", err)
+	}
+
+	v := viper.New()
+	command := &cobra.Command{
+		Use:   "jaeger-ingester",
+		Short: "Jaeger ingester consumes from Kafka and writes to storage",
+		Long:  `Jaeger ingester consumes spans from a particular Kafka topic and writes them to all configured storage types.`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			err := flags.TryLoadConfigFile(v)
+			if err != nil {
+				return err
+			}
+
+			sFlags := new(flags.SharedFlags).InitFromViper(v)
+			logger, err := sFlags.NewLogger(zap.NewProductionConfig())
+			if err != nil {
+				return err
+			}
+			hc, err := sFlags.NewHealthCheck(logger)
+			if err != nil {
+				logger.Fatal("Could not start the health check server.", zap.Error(err))
+			}
+
+			mBldr := new(pMetrics.Builder).InitFromViper(v)
+			baseFactory, err := mBldr.CreateMetricsFactory("jaeger")
+			if err != nil {
+				logger.Fatal("Cannot create metrics factory.", zap.Error(err))
+			}
+			metricsFactory := baseFactory.Namespace("ingester", nil)
+
+			options := app.Options{}
+			options.InitFromViper(v)
+
+			storageFactory.InitFromViper(v)
+			if err := storageFactory.Initialize(baseFactory, logger); err != nil {
+				logger.Fatal("Failed to init storage factory", zap.Error(err))
+			}
+			spanWriter, err := storageFactory.CreateSpanWriter()
+			if err != nil {
+				logger.Fatal("Failed to create span writer", zap.Error(err))
+			}
+			spanConsumer, err := NewSpanConsumer(options, spanWriter, logger, metricsFactory)
+			if err != nil {
+				logger.Fatal("Unable to set up span consumer", zap.Error(err))
+			}
+			spanConsumer.Start()
+
+			hc.Ready()
+			select {
+			case <-signalsChannel:
+				err := spanConsumer.Close()
+				if err != nil {
+					logger.Error("Failed to close consumer", zap.Error(err))
+				}
+				if closer, ok := spanWriter.(io.Closer); ok {
+					err := closer.Close()
+					if err != nil {
+						logger.Error("Failed to close span writer", zap.Error(err))
+					}
+				}
+				logger.Info("Jaeger Ingester is finishing")
```

Comment:
nit: I'd also add a log line on L101 stating "Beginning shutdown of Jaeger Ingester" or something similar.